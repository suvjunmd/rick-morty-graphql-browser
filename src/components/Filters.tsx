import { FormEvent, useEffect } from "react";
import { Form } from "react-router-dom";
import { FilterConfig, getFilterConfig } from "../utils";

interface FiltersProps {
  name?: string;
  status?: string;
  species?: string;
  type?: string;
  gender?: string;
  onSubmit: (config: FilterConfig) => void;
}

export default function Filters({
  name,
  status,
  species,
  type,
  gender,
  onSubmit,
}: FiltersProps) {
  useEffect(() => {
    //restore form values when using the Back button in the browser
    (document.getElementById("name") as HTMLInputElement).value = name || "";
    (document.getElementById("status") as HTMLInputElement).value =
      status || "";
    (document.getElementById("species") as HTMLInputElement).value =
      species || "";
    (document.getElementById("type") as HTMLInputElement).value = type || "";
    (document.getElementById("gender") as HTMLInputElement).value =
      gender || "";
  }, [name, status, species, type, gender]);

  function handleSubmit(e: FormEvent<HTMLFormElement>) {
    e.preventDefault();
    
    // remove empty values from form data
    const filterConfig = getFilterConfig((item) => e.currentTarget[item]?.value);
    
    onSubmit(filterConfig);
  }

  function handleReset() {
    onSubmit({});
  }

  return (
    <div>
      <Form
        id="filter-form"
        role="search"
        className="filter-form"
        onSubmit={handleSubmit}
        onReset={handleReset}
      >
        <fieldset>
          <legend>
            <b>Filters</b>
          </legend>
          <label htmlFor="name">Name</label>
          <br />
          <input id="name" type="search" name="name" defaultValue={name} />
          <br />
          <label htmlFor="status">Status</label>
          <br />
          <select id="status" name="status" defaultValue={status}>
            <option value=""></option>
            <option value="Alive">Alive</option>
            <option value="Dead">Dead</option>
            <option value="unknown">unknown</option>
          </select>
          <br />
          <label htmlFor="species">Species</label>
          <br />
          <input
            id="species"
            type="text"
            name="species"
            defaultValue={species}
          />
          <br />
          <label htmlFor="type">Type</label>
          <br />
          <input id="type" type="text" name="type" defaultValue={type} />
          <br />
          <label htmlFor="gender">Gender</label>
          <br />
          <select id="gender" name="gender" defaultValue={gender}>
            <option value=""></option>
            <option value="Male">Male</option>
            <option value="Female">Female</option>
            <option value="Genderless">Genderless</option>
            <option value="unknown">unknown</option>
          </select>
          <br />
          <br />
          <button type="submit">Filter</button>
          <button type="reset">Reset</button>
        </fieldset>
      </Form>
    </div>
  );
}
