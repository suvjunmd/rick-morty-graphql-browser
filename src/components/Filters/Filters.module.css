.form {
  max-width: fit-content;
  margin-bottom: 1rem;
}
