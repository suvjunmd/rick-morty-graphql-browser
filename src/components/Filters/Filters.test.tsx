import { expect, it } from "vitest";
import { render, screen, fireEvent } from "@testing-library/react";
import Filters from "./Filters";
import { RouterProvider, createMemoryRouter } from "react-router-dom";

it("should render filters", () => {
  const routes = [
    {
      path: "/",
      element: (
        <Filters
          name=""
          status=""
          species=""
          type=""
          gender=""
          onSubmit={() => {}}
        />
      ),
    },
  ];
  const router = createMemoryRouter(routes);
  const { container } = render(<RouterProvider router={router} />);
  expect(screen.getByText("Filters")).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});

it("should call onSubmit prop when clicked on Filter button", () => {
  const handleClick = vi.fn();
  const routes = [
    {
      path: "/",
      element: (
        <Filters
          name=""
          status=""
          species=""
          type=""
          gender=""
          onSubmit={handleClick}
        />
      ),
    },
  ];
  const router = createMemoryRouter(routes);
  render(<RouterProvider router={router} />);
  fireEvent.click(screen.getByText('Filter'));
  expect(handleClick).toHaveBeenCalledTimes(1)
});  

it("should call onSubmit prop when clicked on Reset button", () => {
  const handleClick = vi.fn();
  const routes = [
    {
      path: "/",
      element: (
        <Filters
          name=""
          status=""
          species=""
          type=""
          gender=""
          onSubmit={handleClick}
        />
      ),
    },
  ];
  const router = createMemoryRouter(routes);
  render(<RouterProvider router={router} />);
  fireEvent.click(screen.getByText('Reset'));
  expect(handleClick).toHaveBeenCalledTimes(1)
});  