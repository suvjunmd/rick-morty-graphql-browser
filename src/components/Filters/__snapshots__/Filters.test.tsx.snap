// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render filters 1`] = `
<div>
  <div>
    <form
      action="/"
      class="_form_950627"
      id="filter-form"
      method="get"
      role="search"
    >
      <fieldset>
        <legend>
          <b>
            Filters
          </b>
        </legend>
        <label
          for="name"
        >
          Name
        </label>
        <br />
        <input
          id="name"
          name="name"
          type="search"
          value=""
        />
        <br />
        <label
          for="status"
        >
          Status
        </label>
        <br />
        <select
          id="status"
          name="status"
        >
          <option
            selected=""
            value=""
          />
          <option
            value="Alive"
          >
            Alive
          </option>
          <option
            value="Dead"
          >
            Dead
          </option>
          <option
            value="unknown"
          >
            unknown
          </option>
        </select>
        <br />
        <label
          for="species"
        >
          Species
        </label>
        <br />
        <input
          id="species"
          name="species"
          type="text"
          value=""
        />
        <br />
        <label
          for="type"
        >
          Type
        </label>
        <br />
        <input
          id="type"
          name="type"
          type="text"
          value=""
        />
        <br />
        <label
          for="gender"
        >
          Gender
        </label>
        <br />
        <select
          id="gender"
          name="gender"
        >
          <option
            selected=""
            value=""
          />
          <option
            value="Male"
          >
            Male
          </option>
          <option
            value="Female"
          >
            Female
          </option>
          <option
            value="Genderless"
          >
            Genderless
          </option>
          <option
            value="unknown"
          >
            unknown
          </option>
        </select>
        <br />
        <br />
        <button
          type="submit"
        >
          Filter
        </button>
        <button
          type="reset"
        >
          Reset
        </button>
      </fieldset>
    </form>
  </div>
</div>
`;
